File: 2001-3000/2080. Range Frequency Queries.java
// Medium
// HashMap, Binary Search
// Time:O(log m),Space:O(n)
// https://leetcode.cn/problems/range-frequency-queries/

import java.util.*;

class RangeFreqQuery {
    // Use Map to store (value, its appear indexs)
    // Then use Binary Search to find first and last appear index
    
    private Map<Integer, List<Integer>> map;

    public RangeFreqQuery(int[] arr) {
        // O(n)
        map = new HashMap<>();
        for (int i = 0; i < arr.length; i++) {
            map.putIfAbsent(arr[i], new ArrayList<>());
            map.get(arr[i]).add(i);
        }
    }
    
    public int query(int left, int right, int value) {
        // O(log m)
        if (!map.containsKey(value)) return 0;
        List<Integer> indexs = map.get(value);
        //  - after left / before right
        int start = lowerBound(indexs, left); // find first pos >= target(left)
        int end = upperBound(indexs, right); // find first pos > target(right)
        return Math.max(0, end - start);
    }

    private int lowerBound(List<Integer> list, int target) {
        int left = 0, right = list.size();
        while (left < right) {
            int mid = (left + right) >> 1;
            if (list.get(mid) >= target) right = mid;
            else left = mid + 1;
        }
        return left;
    }

    private int upperBound(List<Integer> list, int target) {
        int left = 0, right = list.size();
        while (left < right) {
            int mid = (left + right) >> 1;
            if (list.get(mid) > target) right = mid;
            else left = mid + 1;
        }
        return left;
    }
}

File: 2001-3000/2080. Range Frequency Queries.py
from typing import List

class RangeFreqQuery:

    def __init__(self, arr: List[int]):
        self.pos = {}
        for i, num in enumerate(arr):
            if num not in self.pos:
                self.pos[num] = []
            self.pos[num].append(i)

    def query(self, left: int, right: int, value: int) -> int:
        if value not in self.pos:
            return 0
        
        indices = self.pos[value]

        # find first >= left（左边界）
        lo, hi = 0, len(indices) - 1
        left_pos = len(indices)
        while lo <= hi:
            mid = (lo + hi) // 2
            if indices[mid] >= left:
                left_pos = mid
                hi = mid - 1
            else:
                lo = mid + 1
        
        if left_pos == len(indices):
            return 0

        # find last <= right (右边界)
        lo, hi = left_pos, len(indices) - 1
        right_pos = -1
        while lo <= hi:
            mid = (lo + hi) // 2
            if indices[mid] <= right:
                right_pos = mid
                lo = mid + 1
            else:
                hi = mid - 1
        
        if right_pos == -1:
            return 0
        
        return right_pos - left_pos + 1
    